### DISCUSSION TITLE
TraceBase File Format Support

### DISCUSSION THREAD

**Discussion by lparsons** [Princeton-LSI-ResearchComputing/tracebase#1039] [Ideas]
TraceBase needs a flexible system to managing input data formats.

There is a wide and expanding range of tools generating peak annotation data. Many of these tools use custom data formats that are not well defined.

The goal of this discussion to decide on an architecture and plan for supporting these various data formats. Once that is done, we will create issues to define specific changes and additions to make.

### Currently Supported Formats

- [AccuCor](https://cran.r-project.org/web/packages/accucor/index.html) / MAVEN1
- [isocorr13C15N](https://github.com/xxing9703/isocorr13C15N)
- [IsoAutocorr](https://github.com/xxing9703/Iso-Autocorr)
- UniCorr (universal tracebase-specific format)

### Formats Requiring Support

- [MAVEN2](https://github.com/eugenemel/maven)
- Xi's Automated Peak Picking MATLAB script - code not avail yet

### Formats with No Support Planned

- [AccuCor2](https://doi.org/10.1038/s41374-021-00631-4)

## Proposal

1. Create a PeakAnnotationDataTable class
2. Write parsers for each file format that read files and create PeakAnnotationDataTable objects
3. All downstream TraceBase code will utilize PeakAnnotationDataTable objects
4. Gather example files for each format, organize in test data directory, and have tests that parse each into a PeakAnnotationDataTable object.

## Other Considerations

- There may exist published data format standards that would be useful to support. However, there are many tools in use and standard formats are not well utilized atm. These formats also tend to be more cumbersome to use in Excel, etc. and require specialized tools.
	- mzTab-M - https://doi.org/10.1021/acs.analchem.8b04310
	- mzQuantML - https://doi.org/10.1074%2Fmcp.O113.028506
	- mzIdentML - https://doi.org/10.1074%2Fmcp.M111.014381

**Comment by hepcat72** [comment]
> 1. Create a PeakAnnotationDataTable class

This exists, BTW.  It's called `PeakAnnotationsLoader`.  You just create derived classes off it that just define the conversion of the various formats to the universal format (`Unicorr`).  If there is any novel type of data in these new formats that we want to utilize, it would require editing to the superclass, but otherwise, any new format just needs a small derived class that defines the conversion.

**Reply by hepcat72** [reply]
> 2. Write parsers for each file format that read files and create PeakAnnotationDataTable objects

Also, technically, the parser is pandas' `read*` methods and the conversion is done using half pandas methods and half python code.  What needs to be written are small derived classes.  The ones that currently exist are:

- `AccucorLoader`
- `IsocorrLoader`
- `IsoautocorrLoader`
- `UnicorrLoader`

They each *just* specify data structures that the superclass needs to do the conversion, however, if any conversion requirement is not supported, you can create a derived class and write a custom converter that overrides the `convert_df()` method.  You can put any code in there you want.  It just has to return the converted (single) dataframe (which matches the Unicorr format).

**Comment by mneinast** [comment]
1) Just want to reiterate that Tracebase is designed to only accept natural-abundance-corrected data (except for the raw mzXML, of course).

2) MAVEN2 output is...complicated.  I will try to explain it here:

There are multiple formats for data output.  I have labeled the example outputs by numbers and attached them here:








When you select the "export to csv" button, a submenu provides several additional options.  Selecting either "Export Selected" or "Export All Groups" will generate a csv similar to MAVEN1 output [1_export all.csv](https://github.com/user-attachments/files/16253877/1_export.all.csv).  In addition, you can select "Isotopes Export".  Then, a new window appears:
![image](https://github.com/user-attachments/assets/fa775716-6a5d-4bd9-81ff-aa581a9cc7a1)

The "Absolute Abundance" generates a table analogous to the "Corrected" tab in Accucor [2_export isotopes_absolute abundance_corrected.csv](https://github.com/user-attachments/files/16253876/2_export.isotopes_absolute.abundance_corrected.csv).  If we decide to support MAVEN2 output, then Tracebase should support this one.  In contrast, the "Fractional Abundance" is missing the abundance information, and we could decide to NOT support that format [3_export isotopes_fractional abundance_corrected.csv](https://github.com/user-attachments/files/16253875/3_export.isotopes_fractional.abundance_corrected.csv).

As seen in the "Isotopes Export", you can output uncorrected or corrected data.  The examples **(2)** and **(3)** are corrected. However, the correction algorithm is more simple than what is implemented by AccuCor and other tools.

In addition to all of this, inside the MAVEN2 results window, you can select "switch between Group and Peak views".  This pivots the data and adds additional columns. The pivoted views includes additional information related to a more complex mass-spec method, "MS2", that we do not routinely implement (yet).  The csv output from "Export All" reflects this [1b_export all_pivoted view.csv](https://github.com/user-attachments/files/16253874/1b_export.all_pivoted.view.csv).  However, if you perform a correction for natural abundance, the exported csv is similar to the other isotope corrected results [2b_export isotopes_absolute abundance_corrected.csv](https://github.com/user-attachments/files/16253872/2b_export.isotopes_absolute.abundance_corrected.csv).

So MAVEN2 can generate multiple corrected outputs.  Most in our lab are not yet using MAVEN2, but I expect more to adopt it over time because it is the only actively maintained software at the moment.

**Comment by mneinast** [comment]
Xi's Automated Peak Picking is under early development, but if we can solidify the output format that Tracebase will use now, then we can start accepting this data and revise it later if necessary.  Xi is using this tool to analyze a lot of newly generated data from new and old experiments, and we should try hard to get this into Tracebase.

It may be nice to distinguish automated peaks from the manually curated peaks (or more generally, distinguish the sources of every input type).

In addition, we could consider (one day) incorporating this automated peak-picking into Tracebase directly.  This could provide a big incentive for researchers to upload to Tracebase, and centralize its implementation.  This of course only makes sense after it has matured, and would require significant work.

**Comment by lparsons** [comment]
Sounds like @hepcat72 has already laid some good groundwork for supporting additional file types. I'm going to create some separate discussions for @xxing9703 's MATLAB software and MAVEN2 support.

* #1056 
* #1057
